pystiche_papers/sanakoyeu_et_al_2018/_utils.py

```diff
@@ -0,0 +1,107 @@
+from typing import Any, Iterator, List, Optional, Union
+
+import torch
+from torch import nn, optim
+from torch.optim.lr_scheduler import ExponentialLR
+from torch.optim.optimizer import Optimizer
+
+from pystiche.ops.container import OperatorContainer
+from pystiche.ops.op import ComparisonOperator
+from pystiche.optim.meter import FloatMeter
+
+
+def optimizer(
+    params: Union[nn.Module, Iterator[torch.nn.parameter.Parameter]]
+) -> Optimizer:
+    r"""Optimizer from :cite:`SKL+2018`.
+
+    Args:
+        params: Parameter to be optimized.
+
+    """
+    if isinstance(params, nn.Module):
+        params = params.parameters()
+    return optim.Adam(params, lr=2e-4)
+
+
+def lr_scheduler(optimizer: Optimizer) -> Optional[ExponentialLR]:
+    r"""Learning rate scheduler from :cite:`SKL+2018`.
+
+    Args:
+        optimizer: Wrapped optimizer.
+
+    """
+    return DelayedExponentialLR(optimizer, gamma=0.1, delay=2)
+
+
+class ExponentialMovingAverage(FloatMeter):
+    def __init__(
+        self,
+        name: str,
+        init_val: float = 0.8,
```

Comment:
Why does this start at such an odd value? `FloatMeter` starts at `0.0`.

Reply:
This `ExponentialMovingAverageMeter` is used to track the `discriminator_success`, which is initialized with `0.8`. Since this `ExponentialMovingAverageMeter` is not used in any other paper so far, I set the initial value to `0.8`.

Reply:
I wouldn't do that. The name does not imply that it is specifically used in this paper and thus the defaults should be general.

Reply:
However, the Exponential Moving Average needs a starting point. Since this calculates the exponential average. So either the first value would have to be set as default or like here a user defined value. So I would suggest to leave out the default value here, so that a value has to be specified when using the `ExponentialMovingAverageMeter`.